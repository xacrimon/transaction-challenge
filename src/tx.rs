const AMOUNT_SHIFT_ACCURACY: f32 = 10000.0;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use std::ops;

#[derive(Deserialize, Clone, Copy)]
pub struct Tx {
    #[serde(rename = "type")]
    pub ty: TxType,

    #[serde(rename = "client")]
    pub client: Client,

    #[serde(rename = "tx")]
    pub id: TxId,

    #[serde(rename = "amount", deserialize_with = "Amount::deserialize", default)]
    pub amount: Amount,
}

#[derive(Deserialize, Clone, Copy)]
pub enum TxType {
    #[serde(rename = "deposit")]
    Deposit,

    #[serde(rename = "withdrawal")]
    Withdrawal,

    #[serde(rename = "dispute")]
    Dispute,

    #[serde(rename = "resolve")]
    Resolve,

    #[serde(rename = "chargeback")]
    Chargeback,
}

#[derive(Deserialize, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
pub struct Client(u16);

impl From<u16> for Client {
    fn from(id: u16) -> Self {
        Client(id)
    }
}

impl Into<u16> for Client {
    fn into(self) -> u16 {
        self.0
    }
}

#[derive(Deserialize, Clone, Copy, Debug, Hash, PartialEq, Eq)]
pub struct TxId(u32);

impl From<u32> for TxId {
    fn from(id: u32) -> Self {
        TxId(id)
    }
}

impl Into<u32> for TxId {
    fn into(self) -> u32 {
        self.0
    }
}

#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Default)]
pub struct Amount(i32);

impl Amount {
    fn deserialize<'de, D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let amount: Option<f32> = Deserialize::deserialize(deserializer)?;
        Ok(Amount::from(amount.unwrap_or(0.0)))
    }

    pub fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        serializer.serialize_f32(self.clone().into())
    }
}

impl From<f32> for Amount {
    fn from(amount: f32) -> Self {
        Amount((amount * AMOUNT_SHIFT_ACCURACY) as i32)
    }
}

impl Into<f32> for Amount {
    fn into(self) -> f32 {
        self.0 as f32 / AMOUNT_SHIFT_ACCURACY
    }
}

impl ops::Add for Amount {
    type Output = Self;

    fn add(self, other: Self) -> Self {
        Amount(self.0 + other.0)
    }
}

impl ops::Sub for Amount {
    type Output = Self;

    fn sub(self, other: Self) -> Self {
        Amount(self.0 - other.0)
    }
}
