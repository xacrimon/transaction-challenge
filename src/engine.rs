use super::tx::{Amount, Client, Tx, TxId, TxType};
use anyhow::{anyhow, Result};
use serde::Serialize;
use std::collections::{HashMap, HashSet};

#[derive(Clone, Copy, Serialize)]
pub struct Account {
    client: Client,
    available: Amount,
    held: Amount,
    total: Amount,
    locked: bool,
}

pub struct Engine {
    accounts: HashMap<Client, Account>,
    tx_amounts: HashMap<TxId, Amount>,
    disputed: HashSet<TxId>,
}

impl Engine {
    pub fn new() -> Self {
        Engine {
            accounts: HashMap::new(),
            tx_amounts: HashMap::new(),
            disputed: HashSet::new(),
        }
    }

    pub fn apply(&mut self, tx: Tx) -> Result<()> {
        self.tx_amounts.insert(tx.id, tx.amount);

        match tx.ty {
            TxType::Deposit => self.deposit(tx.client, tx.amount, tx.id),
            TxType::Withdrawal => self.withdraw(tx.client, tx.amount, tx.id),
            TxType::Dispute => self.dispute(tx.client, tx.amount, tx.id),
            TxType::Resolve => self.resolve(tx.client, tx.amount, tx.id),
            TxType::Chargeback => self.chargeback(tx.client, tx.amount, tx.id),
        }
    }

    fn account(&mut self, client: Client) -> Result<&mut Account> {
        let account = self.accounts.entry(client).or_insert(Account {
            client,
            available: Amount::from(0.0),
            held: Amount::from(0.0),
            total: Amount::from(0.0),
            locked: false,
        });

        if account.locked {
            return Err(anyhow!("account is locked"));
        }

        Ok(account)
    }

    fn deposit(&mut self, client: Client, amount: Amount, _tx_id: TxId) -> Result<()> {
        let account = self.account(client)?;

        account.available = account.available + amount;
        account.total = account.total + amount;
        Ok(())
    }

    fn withdraw(&mut self, client: Client, amount: Amount, tx_id: TxId) -> Result<()> {
        let account = self.account(client)?;

        if account.available < amount {
            return Err(anyhow!(
                "insufficient funds for withdrawal, tx: {:?}",
                tx_id
            ));
        }

        account.available = account.available - amount;
        account.total = account.total - amount;
        Ok(())
    }

    fn dispute(&mut self, client: Client, _amount: Amount, tx_id: TxId) -> Result<()> {
        let amount = *self
            .tx_amounts
            .get(&tx_id)
            .ok_or(anyhow!("tx not found, tx: {:?}", tx_id))?;
        let account = self.account(client)?;

        if account.available < amount {
            account.locked = true;
            return Err(anyhow!(
                "insufficient funds for dispute, locking account, tx: {:?}",
                tx_id
            ));
        }

        account.available = account.available - amount;
        account.held = account.held + amount;
        Ok(())
    }

    fn resolve(&mut self, client: Client, _amount: Amount, tx_id: TxId) -> Result<()> {
        let amount = *self
            .tx_amounts
            .get(&tx_id)
            .ok_or(anyhow!("tx not found, tx: {:?}", tx_id))?;
        let account = self.account(client)?;

        if account.held < amount {
            return Err(anyhow!(
                "insufficient held funds, this shouldn't happen, tx: {:?}",
                tx_id
            ));
        }

        account.available = account.available + amount;
        account.held = account.held - amount;
        Ok(())
    }

    fn chargeback(&mut self, client: Client, _amount: Amount, tx_id: TxId) -> Result<()> {
        let amount = *self
            .tx_amounts
            .get(&tx_id)
            .ok_or(anyhow!("tx not found, tx: {:?}", tx_id))?;
        let account = self.account(client)?;

        if account.held < amount {
            return Err(anyhow!(
                "insufficient held funds, this shouldn't happen, tx: {:?}",
                tx_id
            ));
        }

        account.held = account.held - amount;
        account.total = account.total - amount;
        account.locked = true;
        Ok(())
    }

    pub fn accounts(&self) -> impl Iterator<Item = Account> + '_ {
        self.accounts.values().copied()
    }
}

// ignore resolve, chargeback in some cases
