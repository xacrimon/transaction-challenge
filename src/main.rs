mod engine;
mod tx;

use anyhow::Result;
use clap::Parser;
use engine::Engine;
use std::fs::File;
use std::io::{self, Read};
use std::path::PathBuf;
use tx::Tx;

fn stream_records_from_reader<'r>(
    reader: &'r mut dyn Read,
) -> impl Iterator<Item = Result<Tx>> + 'r {
    csv::Reader::from_reader(reader)
        .into_deserialize()
        .map(|r| r.map_err(Into::into))
}

#[derive(Parser, Debug)]
#[clap(author, version, about, long_about = None)]
struct Args {
    file: PathBuf,
}

fn main() -> Result<()> {
    let args = Args::parse();
    let mut file = File::open(&args.file)?;
    let tx_stream = stream_records_from_reader(&mut file);
    let mut writer = csv::Writer::from_writer(io::stdout());
    let mut engine = Engine::new();

    for tx in tx_stream {
        if let Err(err) = engine.apply(tx?) {
            eprintln!("{}", err);
        }
    }

    for account in engine.accounts() {
        writer.serialize(account)?;
    }

    writer.flush()?;
    Ok(())
}

#[cfg(test)]
mod tests {
    use super::stream_records_from_reader;
    use super::Engine;
    use std::fs::File;

    fn verify_processing() {
        let mut file = File::open("sample.csv").unwrap();
        let tx_stream = stream_records_from_reader(&mut file);
        let mut data = Vec::new();
        let mut wtr = std::io::Cursor::new(&mut data);
        let mut writer = csv::Writer::from_writer(&mut wtr);
        let mut engine = Engine::new();

        for tx in tx_stream {
            engine.apply(tx.unwrap()).unwrap();
        }

        for account in engine.accounts() {
            writer.serialize(account).unwrap();
        }

        writer.flush().unwrap();
        drop(writer);
        
        let readable = String::from_utf8(data).unwrap();
        insta::assert_snapshot!(readable);
    }
}
