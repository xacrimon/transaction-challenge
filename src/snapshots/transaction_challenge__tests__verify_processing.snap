---
source: src/main.rs
expression: readable
---
client,available,held,total,locked
client,available,held,total,locked
1,2.0,0.0,2.0,false
2,2.0,0.0,2.0,false
3,0.0,5.0,5.0,false
4,10.5562,0.0,10.5562,false

